File: services/voice/internal/s3/s3.go
// Package s3 provides upload, presigned URL, and delete helpers
// for Backblaze B2 using the S3-compatible API (endpoint: s3.us-west-004.backblazeb2.com).
package s3

import (
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	voiceconfig "github.com/go-grpc-sqlc/voice/config"
)

// Client wraps the AWS S3 client configured for Backblaze B2.
type Client struct {
	inner  *s3.Client
	bucket string
}

// New creates a Backblaze B2 S3 client from the voice service config.
// Backblaze B2 requires path-style addressing and a custom endpoint.
func New(cfg voiceconfig.Config) (*Client, error) {
	awsCfg, err := awsconfig.LoadDefaultConfig(context.Background(),
		awsconfig.WithRegion(cfg.S3Region),
		awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			cfg.S3AccessKey,
			cfg.S3SecretKey,
			"",
		)),
	)
	if err != nil {
		return nil, fmt.Errorf("s3: load aws config: %w", err)
	}

	// Backblaze B2 needs path-style URLs and a fixed base endpoint.
	inner := s3.NewFromConfig(awsCfg, func(o *s3.Options) {
		o.BaseEndpoint = aws.String(cfg.S3Endpoint)
		o.UsePathStyle = true
	})

	return &Client{inner: inner, bucket: cfg.S3Bucket}, nil
}

// UploadOptions configures an object upload.
type UploadOptions struct {
	Key         string
	Body        []byte
	ContentType string // defaults to "audio/wav"
}

// Upload puts an object into the Backblaze B2 bucket.
func (c *Client) Upload(ctx context.Context, opts UploadOptions) error {
	if opts.ContentType == "" {
		opts.ContentType = "audio/wav"
	}
	_, err := c.inner.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(c.bucket),
		Key:         aws.String(opts.Key),
		Body:        bytesReader(opts.Body),
		ContentType: aws.String(opts.ContentType),
	})
	if err != nil {
		return fmt.Errorf("s3: upload %q: %w", opts.Key, err)
	}
	return nil
}

// Delete removes an object from the Backblaze B2 bucket.
func (c *Client) Delete(ctx context.Context, key string) error {
	_, err := c.inner.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(c.bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return fmt.Errorf("s3: delete %q: %w", key, err)
	}
	return nil
}

// GetSignedURL returns a presigned GET URL valid for 1 hour.
func (c *Client) GetSignedURL(ctx context.Context, key string) (string, error) {
	presigner := s3.NewPresignClient(c.inner)
	req, err := presigner.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(c.bucket),
		Key:    aws.String(key),
	}, s3.WithPresignExpires(time.Hour))
	if err != nil {
		return "", fmt.Errorf("s3: presign %q: %w", key, err)
	}
	return req.URL, nil
}

File: web/src/lib/s3.ts
import "server-only"

import {
  DeleteObjectCommand,
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3"
import { getSignedUrl } from "@aws-sdk/s3-request-presigner"

const s3Endpoint = process.env.S3_ENDPOINT
const s3Region = process.env.S3_REGION
const s3Bucket = process.env.S3_BUCKET
const s3AccessKey = process.env.S3_ACCESS_KEY_ID
const s3SecretKey = process.env.S3_SECRET_ACCESS_KEY

if (!s3Endpoint || !s3Region || !s3Bucket || !s3AccessKey || !s3SecretKey) {
  throw new Error("Missing Backblaze B2 S3 environment variables")
}

const b2 = new S3Client({
  region: s3Region,
  endpoint: s3Endpoint,
  forcePathStyle: true,
  credentials: {
    accessKeyId: s3AccessKey,
    secretAccessKey: s3SecretKey,
  },
})

type UploadAudioOptions = {
  buffer: Buffer
  key: string
  contentType?: string
}

export async function uploadAudio({
  buffer,
  key,
  contentType = "audio/wav",
}: UploadAudioOptions): Promise<void> {
  await b2.send(
    new PutObjectCommand({
      Bucket: s3Bucket,
      Key: key,
      Body: buffer,
      ContentType: contentType,
    }),
  )
}

export async function deleteAudio(key: string): Promise<void> {
  await b2.send(
    new DeleteObjectCommand({
      Bucket: s3Bucket,
      Key: key,
    }),
  )
}

export async function getSignedAudioUrl(key: string): Promise<string> {
  const command = new GetObjectCommand({
    Bucket: s3Bucket,
    Key: key,
  })

  return getSignedUrl(b2, command, { expiresIn: 3600 })
}
